### PR TITLE
Add tensor factory `torch.tensor`

### PR DESCRIPTION
**Pull request by zasdfgbnm**
[`torch.tensor`](https://pytorch.org/docs/stable/generated/torch.tensor.html?highlight=tensor#torch.tensor) is a tensor factory that constructs a tensor from given scalar values. This PR adds support for this tensor factory to our C++ API. Python frontend is not supported yet, because of the following reason:

This tensor factory is not something we prioritize to have. However, building this out requires improving our array support in codegen. The main purpose of this PR is to build out array infrastructure and a great portion of this work will be reused for TMA support (which takes an argument of N dimensional coordinates as index). The prototype for TMA is at https://github.com/NVIDIA/Fuser/pull/455

In the future, if this tensor factory turns out to be useful, we can add python frontend at that time. (It will NOT be easy, because python's `[[1, 2], [2, 4]]` and `[1, 2]` are identical types, but in C++, they are different types)

Generated code for `TensorFactoryTest.TensorConstruct`:

```CUDA
__global__ void kernel1(int64_t i0, int64_t i1, int64_t i2, int64_t i3, Tensor<int64_t, 2, 2> T0) {
  NVFUSER_DEFINE_MAGIC_ZERO;
  Array<int64_t, 2, 1> a4;
  a4 = {i2, i3};
  Array<int64_t, 2, 1> a5;
  a5 = {i0, i1};
  Array<Array<int64_t, 2, 1>, 2, 1> a6;
  a6 = {a5, a4};
  #pragma unroll
  for(nvfuser_index_t i7 = 0; i7 < 2; ++i7) {
    int64_t i8;
    i8 = 2 * i7;
    Array<int64_t, 2, 1> a9;
    a9 = a6[i7];
    #pragma unroll
    for(nvfuser_index_t i10 = 0; i10 < 2; ++i10) {
      int64_t i11;
      i11 = i10 + nvfuser_zero;
      T0[(i8 + i11)] = (a9[i11]);
    }
  }
  NVFUSER_UPDATE_MAGIC_ZERO;
}
```

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
BTW, I don't think `a9` will use register due to `nvfuser_zero`
```C++
      i11 = i10 + nvfuser_zero;
      T0[(i8 + i11)] = (a9[i11]);
```
But this PR only cares about correctness.

File: csrc/dispatch.h [hunk 1]
```diff
@@ -165,13 +168,18 @@ class TORCH_CUDA_CU_API OptOutConstDispatch : public PolymorphicBase {
 
   virtual void handle(const AggregateVal*);
 
+  virtual void handleUntypedVal(const Val*);
```
Comment (zasdfgbnm) [inline csrc/dispatch.h:171]: I think we should clean this up by making `Scalar` not a template class, and uses `std::any` to store its value. But I will not work on it in this PR.

**Comment by zasdfgbnm** [comment]
!build

File: csrc/dispatch.h (hunk 1 above)
Comment (zasdfgbnm) [reply csrc/dispatch.h:171]: Oh, no, `std::any` do not have `operator==`....

**Comment by zasdfgbnm** [comment]
!build

File: csrc/ops/arith.h [hunk 2]
```diff
@@ -770,4 +771,13 @@ TORCH_CUDA_CU_API TensorView* fusedMultiplySum(
     const std::vector<int>& axes,
     Val* init = nullptr);
 
+// Create a tensor view from the given value. The given value can be a single
+// scalar, an array of scalars, or a nested array of scalars.
+TensorView* tensor(Val* val);
```
Comment (naoyam) [inline csrc/ops/arith.h:776]: So, if it's a scalar, it will create a 0-dim tensor and the scalar value is used to initialize the 0-dim tensor. If it's an array, the value is used to determine the extent of each dimension. Is this correct?

File: csrc/ir/builder.cpp [hunk 3]
```diff
@@ -21,6 +21,9 @@ Val* IrBuilder::newScalar(DataType dtype) {
   if (isPointerType(dtype)) {
     return IrBuilder::create<Int>(dtype);
   }
+  if (!std::holds_alternative<PrimDataType>(dtype.type)) {
```
Comment (naoyam) [inline csrc/ir/builder.cpp:24]: What case is this supposed to handle? I guess we are creating a scalar of a non-primitive data type?

File: csrc/ops/arith.h (hunk 2 above)
Comment (zasdfgbnm) [reply csrc/ops/arith.h:776]: Yes, for example if you have `val` with data type `Array<Array<Float, 5>, 3>`, then you will have a float tensor with shape `(5, 3)`.

File: csrc/ir/builder.cpp (hunk 3 above)
Comment (zasdfgbnm) [reply csrc/ir/builder.cpp:24]: Yes, we are creating scalar with data type `ArrayOf`

File: csrc/dispatch.h (hunk 1 above)
Comment (jacobhinkle) [reply csrc/dispatch.h:171]: Since there are only a handful of scalar types, could we use `std::variant` instead of `std::any`?

**Comment by jacobhinkle** [comment]
As an alternative to new scalar machinery, could you instead specify a vector of scalars along with a shape? Or were you thinking of creating new scalar ops to replicate some tensor ops?

File: csrc/dispatch.h (hunk 1 above)
Comment (zasdfgbnm) [reply csrc/dispatch.h:171]: > Since there are only a handful of scalar types, could we use `std::variant` instead of `std::any`?

No, we can have unlimited number of types:
```
Array<float, 3>
Array<float, 4>
Array<Array<float, 3>, 4>
...
```

**Comment by zasdfgbnm** [comment]
> As an alternative to new scalar machinery, could you instead specify a vector of scalars along with a shape? 

What is the benefit? It is not consistent with torch API.

> Or were you thinking of creating new scalar ops to replicate some tensor ops?

I am not trying to replicate tensor ops. I just feel this is the most natural way to do.

File: csrc/dispatch.cpp [hunk 4]
```diff
@@ -52,6 +52,10 @@ void Val::dispatch(T handler, Val* val) {
         ptr(handler)->handle(val->as<Int>());
         return;
       }
+      if (!std::holds_alternative<PrimDataType>(val->getDataType()->type)) {
+        ptr(handler)->handleUntypedVal(val);
```
Comment (naoyam) [inline csrc/dispatch.cpp:56]: Wonder why this is called "untyped". Actually, what "untyped" means?

Comment (zasdfgbnm) [reply csrc/dispatch.cpp:56]: "untyped" means it is not an `Scalar<SomeType>` , it is just a plain `Val` without much detailed information.

Comment (naoyam) [reply csrc/dispatch.cpp:56]: Is it at the moment always an array?

Comment (zasdfgbnm) [reply csrc/dispatch.cpp:56]: Yes, but array of 3 elements and array of 3 elements are different types. And also we can have array of array. So it is actually infinite amount of types.

Comment (zasdfgbnm) [reply csrc/dispatch.cpp:56]: yeah, sure

Comment (zasdfgbnm) [reply csrc/dispatch.cpp]: changed

**Comment by zasdfgbnm** [comment]
!build

File: csrc/dispatch.h [hunk 5]
```diff
@@ -366,6 +379,8 @@ class TORCH_CUDA_CU_API OptOutMutator : public PolymorphicBase {
   virtual void mutate(kir::Predicate*);
   virtual void mutate(kir::TensorIndex*);
 
+  virtual void mutateUntypedVal(Val*);
```
Comment (naoyam) [inline csrc/dispatch.h:382]:
```suggestion
  virtual void mutateArrayVal(Val*);
```

**Review by naoyam — APPROVED** [review]
LGTM. Nice work!

**Comment by zasdfgbnm** [comment]
!build

File: csrc/dispatch.h (hunk 5 above)
Comment (zasdfgbnm) [reply csrc/dispatch.h]: fixed
